cloudnet-modules/cloudnet-cloudperms/src/main/java/de/dytanic/cloudnet/ext/cloudperms/bukkit/BukkitCloudNetCloudPermissionsPermissible.java

```diff
@@ -81,44 +83,55 @@ public boolean hasPermission(@NotNull String inName) {
             if (permissionUser == null) {
                 return false;
             }
-            if (this.checkPermission(permissionUser, inName)) {
-                return true;
+
+            PermissionCheckResult result = this.permissionsManagement.getPermissionResult(permissionUser, inName);
+            if (result != PermissionCheckResult.DENIED) {
+                return result.asBoolean();
             }
-            return this.testParents(inName, parentPermission -> this.checkPermission(permissionUser, parentPermission.getName()));
+
+            return this.testParents(inName, perm -> this.permissionsManagement.getPermissionResult(permissionUser, perm.getName()));
         } catch (Exception ex) {
             ex.printStackTrace();
             return false;
         }
     }
 
-    private boolean testDefaultPermission(IPermissionUser permissionUser, String name) {
-        return this.permissionsManagement.getPermissionResult(permissionUser, name) != PermissionCheckResult.FORBIDDEN;
-    }
-
-    private boolean testParents(String inName, Predicate<Permission> parentAcceptor) {
+    private boolean testParents(String inName, Function<Permission, PermissionCheckResult> parentAcceptor) {
         for (Permission parent : this.player.getServer().getPluginManager().getPermissions()) {
-            if (this.testParents(inName, parent, null, parentAcceptor)) {
-                return true;
+            PermissionCheckResult result = this.testParents(inName, parent, null, parentAcceptor);
+            if (result != PermissionCheckResult.DENIED) {
+                return result.asBoolean();
             }
         }
         return false;
     }
 
-    private boolean testParents(String inName, Permission parent, Permission lastParent, Predicate<Permission> parentAcceptor) {
+    private PermissionCheckResult testParents(String inName, Permission parent, Permission lastParent, Function<Permission, PermissionCheckResult> parentAcceptor) {
         for (Map.Entry<String, Boolean> entry : parent.getChildren().entrySet()) {
```

Comment:
make a better conndition for the for loop, for ex. use a iterator. Prevent a hard break with return